### PR TITLE
add isRoot and support it in build_runner

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/build/issues/1334

cc @grouma

### CODE REVIEW

File: build/lib/src/builder/builder.dart [hunk 1]
```diff
@@ -34,18 +34,24 @@ class BuilderOptions {
...
   ///
-  /// Values are overridden at a per-key granularity. There is no value level
-  /// merging. [other] may be null or empty, in which case this instance is
-  /// returned directly.
-  BuilderOptions overrideWith(BuilderOptions other) =>
-      other == null || other.config.isEmpty
-          ? this
-          : new BuilderOptions({}..addAll(config)..addAll(other.config));
+  /// Config values are overridden at a per-key granularity. There is no value
+  /// level merging. [other] may be null or empty, in which case this instance
```
Comment (natebosch) [inline build/lib/src/builder/builder.dart]: no longer returned directly if other is empty. (Also the meaning of "empty" might not make sense anymore)

do we want to do any other checking for the case where `other.config` is empty and `other.isRoot` is already the same as `this.isRoot`?

File: build_runner/lib/src/package_graph/apply_builders.dart [hunk 2]
```diff
@@ -189,16 +197,20 @@ class BuilderApplication {
...
+    var phaseFactory = (PackageNode package, BuilderOptions options,
         InputSet targetSources, InputSet generateFor, bool isReleaseBuild) {
       generateFor ??= defaultGenerateFor;
 
       final optionsWithDefaults = (defaultOptions ?? BuilderOptions.empty)
           .overrideWith(
               isReleaseBuild ? defaultReleaseOptions : defaultDevOptions)
           .overrideWith(options);
+      if (package.isRoot) {
+        optionsWithDefaults.overrideWith(new BuilderOptions({}, isRoot: true));
```
Comment (natebosch) [inline build_runner/lib/src/package_graph/apply_builders.dart]: I think we need `optionsWithDefaults = optionsWithDefaults.overrideWith`...

Comment (jakemac53) [reply build_runner/lib/src/package_graph/apply_builders.dart]: yep.... good catch

File: build/lib/src/builder/builder.dart (hunk 1 above)
Comment (jakemac53) [reply build/lib/src/builder/builder.dart]: I don't think it is worth optimizing really, these objects are cheap

File: build/test/builder/builder_options_test.dart [hunk 3]
```diff
@@ -13,11 +13,26 @@ void main() {
...
       expect(overridden.config,
           {'foo': 'bar', 'baz': 'different', 'more': 'added'});
+      expect(overridden.isRoot, isFalse);
     });
 
-    test('changes nothing when overriding with empty options', () {
+    test('overrides isRoot', () {
+      var defaults = const BuilderOptions(const {}, isRoot: false);
+      var overridden =
+          defaults.overrideWith(const BuilderOptions(const {}, isRoot: true));
```
Comment (natebosch) [inline build/test/builder/builder_options_test.dart]: what do you think of making this const instance a static member of the class?

`defaults.overrideWith(BuilderOptions.forRoot);`

Comment (jakemac53) [reply build/test/builder/builder_options_test.dart]: sure, why not

**Pull request merged**